Add IQS009 rule (#2373)
home-assistant/developers.home-assistant: docs/core/integration-quality-scale/rules/iqs009.md

                self._unavailable_warning_logged = False
```

### Example for integrations using the coordinator

In this example, we have an integration that uses a coordinator to fetch data.
The coordinator has the logic for logging once built in.
The only thing that you need to do in the coordinator is to raise `UpdateFailed` when the device or service is unavailable.

`coordinator.py`
```python
class MyCoordinator(DataUpdateCoordinator[MyData]):
    """Class to manage fetching data."""

    def __init__(self, hass: HomeAssistant, client: MyClient) -> None:
        """Initialize coordinator."""
        super().__init__(
            hass,
            logger=LOGGER,
            name=DOMAIN,
            update_interval=timedelta(minutes=1),
        )
        self.client = client
    
    async def _async_update_data(self) -> MyData:
        try:
            return await self.client.get_data()
        except MyException as ex:
            raise UpdateFailed(f"The device is unavailable: {ex}")
```

## Additional resources

For more information about managing integration state, see the [documentation](../../../integration_fetching_data)

## Exceptions

> There are no exceptions to this rule.

Author: I am wondering if there are exceptions for this rule. I can think of push based integrations that have trouble keeping track of actually being connected to the device

Reviewer: I don't think that matter much, if they can't track it, this rule doesn't apply as there is nothing to log...

### Example for integrations using the coordinator

In this example, we have an integration that uses a coordinator to fetch data.
The coordinator has the logic for logging once built in.
The only thing that you need to do in the coordinator is to raise `UpdateFailed` when the device or service is unavailable.

`coordinator.py`
```python
class MyCoordinator(DataUpdateCoordinator[MyData]):
    """Class to manage fetching data."""

    def __init__(self, hass: HomeAssistant, client: MyClient) -> None:
        """Initialize coordinator."""
        super().__init__(
            hass,
            logger=LOGGER,
            name=DOMAIN,
            update_interval=timedelta(minutes=1),
        )
        self.client = client
    
    async def _async_update_data(self) -> MyData:
        try:
            return await self.client.get_data()
        except MyException as ex:
            raise UpdateFailed(f"The device is unavailable: {ex}")
```

### Example for entities updating via `async_update`

In this example, we have a sensor that updates its value via `async_update`.

> We will log when the sensor is unavailable and log when the sensor is back online.

Reviewer 2: ```suggestion
The example will log when the sensor is unavailable and log when the sensor is back online.
```

---

> title: "IQS009 - If internet/device/service unavailable, log once unavailable and once when back connected"

Reviewer 2: ```suggestion
title: "IQS009 - If internet/device/service is unavailable, log once when unavailable and once when back connected"
```

---

IQS009 is as follows:

> If internet/device/service unavailable, log once unavailable and once when back connected

Reviewer 2: ```suggestion
If internet/device/service is unavailable, log once when unavailable and once when back connected.
```

## Reasoning

When a device or service is not reachable, the entities will usually go to unavailable.

> To allow the user to find out why this is happening, we should log when this happens.

Reviewer 2: ```suggestion
To allow the user to find out why this is happening, the integration should log when this happens.
```

Be sure to log only once in total to avoid spamming the logs.

> When the device or service is reachable again, we should log that as well.

Reviewer 2: ```suggestion
When the device or service is reachable again, the integration should log that as well.
```

            raise UpdateFailed(f"The device is unavailable: {ex}")
```

### Example for entities updating via `async_update`

In this example, we have a sensor that updates its value via `async_update`.
We will log when the sensor is unavailable and log when the sensor is back online.

> Do note that we keep track of the warning being logged to avoid spamming the logs.

Reviewer 2: ```suggestion
Note that an instance attribute is used to track if the message has been logged to avoid spamming the logs.
```
